sstables/sstables.cc

```diff
@@ -1471,7 +1471,10 @@ sstable::load_owner_shards(const dht::sharder& sharder) {
     // If sharding metadata is not available, we must load first and last keys from summary
     // for sstable::compute_shards_for_this_sstable() to operate on them.
     if (!has_valid_sharding_metadata) {
-        sstlog.warn("Sharding metadata not available for {}, so Summary will be read to allow Scylla to compute shards owning the SSTable.", get_filename());
+        if (_schema->try_get_static_sharder() != nullptr) {
+            // The table uses vnodes but sharding metadata is not available
+            sstlog.warn("Sharding metadata not available for {}, so Summary will be read to allow Scylla to compute shards owning the SSTable.", get_filename());
+        }
         co_await read_summary();
```

Comment:
Should skip reading the summary too (and adjust the title to reflect that, it's more important than the warning).

This can improve startup times.

Reply:
> We need summary for first and last keys.

I thought we had separate storage for first/last key. Maybe that's only in memory and we don't store it separately in some O(1) component?


> What we can do is to optimize it by skipping directly to the location where keys are stored

Well, we need to read the summary anyway. Do we have good prediction about the owner shard? I don't want the summary to end up in non-local memory because of this.

If we end up with non-local memory, perhaps we should copy the summary once we find the right shard.

Reply:
> > We need summary for first and last keys.
> 
> I thought we had separate storage for first/last key. Maybe that's only in memory and we don't store it separately in some O(1) component?

I am thinking we should just insert first and last key in sharding metadata. That would be our O(1) component.

> 
> > What we can do is to optimize it by skipping directly to the location where keys are stored
> 
> Well, we need to read the summary anyway. Do we have good prediction about the owner shard? I don't want the summary to end up in non-local memory because of this.
> 
> If we end up with non-local memory, perhaps we should copy the summary once we find the right shard.

We would end up discarding summary, today. load_owner_shards() is supposed to be a quick way to determine sstable ownership, so sstable components can be loaded locally to the shard that owns it. We either skip to keys, or just have sharding metadata store first and last key. I like the latter more, as it's the same code path for both vnode and tablets,. And we wouldn't be introducing a new mechanism.

Reply:
Reading summary is only there for compatibility with sstables not generated by scylla. But we want sharding_metadata available, so load_owner_shards() can be efficient. It doesn't even read other sub-components in scylla.db file.

Reply:
Why do we need to go through `compute_shards_for_this_sstable` for sstables in tablets-based tables?

Reply:
> Why do we need to go through `compute_shards_for_this_sstable` for sstables in tablets-based tables?

Because it's the generic code for determining ownership using sharder, which is implemented for tablets too.

Reply:
isn't it bogus for tablets?
the owning shard for tablets is determine by the tablets map.

I'm not sure yet if it's related, but we're seeing a crash loop in SCT in a multi-dc, multi-rack setup, where right after a burst of `Sharding metadata not available` the starting node runs out of memory when trying to allocate 128KB in
```
seastar::memory::on_allocation_failure(unsigned long) at ./build/release/seastar.lto/./seastar/src/core/memory.cc:2135
seastar::memory::finish_allocation(void*, unsigned long) at ./build/release/seastar.lto/./seastar/src/core/memory.cc:1627
 (inlined by) seastar::memory::allocate_slowpath(unsigned long) at ./build/release/seastar.lto/./seastar/src/core/memory.cc:1662
seastar::memory::allocate(unsigned long) at ././seastar/src/core/memory.cc:1679
 (inlined by) malloc at ././seastar/src/core/memory.cc:2184
 (inlined by) utils::chunked_vector<std::pair<sstables::metadata_type, unsigned int>, 131072ul>::new_chunk(unsigned long) at ././utils/chunked_vector.hh:383
 (inlined by) utils::chunked_vector<std::pair<sstables::metadata_type, unsigned int>, 131072ul>::make_room(unsigned long, bool) at ././utils/chunked_vector.hh:430
 (inlined by) utils::chunked_vector<std::pair<sstables::metadata_type, unsigned int>, 131072ul>::reserve(unsigned long) at ././utils/chunked_vector.hh:171
 (inlined by) seastar::future<void> sstables::parse<unsigned int, std::pair<sstables::metadata_type, unsigned int> >(schema const&, sstables::sstable_version_types, sstables::random_access_reader&, sstables::disk_array<unsigned int, std::pair<sstables::metadata_type, unsigned int> >&) at ./sstables/sstables.cc:352
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/coroutine:240
```

Reply:
See https://argus.scylladb.com/workspace?state=WyJhZjVkYmYyOS0wMjYwLTRmNzktOGNjYS1hZDVhZWUyMWVjM2EiXQ